### ISSUE TITLE
Why do we need parent identities on frontend?

### ISSUE THREAD

**Issue by madonzy** [magento/magento2#35546]
I've noticed that we have plugin `\Magento\ConfigurableProduct\Model\Plugin\ProductIdentitiesExtender::afterGetIdentities` that is responsible to add parent tags for each child product. For example, that makes the product listing page much slower because we need to load the parent for each product on the page.

Why do we really need that on frontend? I tested and it PURGE cache for all child products when configurable product is saved.

**Comment by engcom-Hotel** [comment]
Hello @madonzy,

Thanks for the reporting and collaboration!

We request you to more elaborate on the issue so that we can reproduce it. Please provide us the below information:

- Environment information
- Steps to reproduce 
- Expected and Actual result
- In case a specific use case, please provide those also
- Any screencast or screenshot

Thanks

**Comment by hostep** [comment]
Quoting from [@sivaschenko's remark in another thread](https://github.com/magento/magento2/issues/6401#issuecomment-602226975):

> child products affect parent product representation while parent products do not affect child products representation on the frontend, so in order to be sure that the cache is valid - the cache of pages with parent products should be invalidated when a child product is changed.

For example, when a child product becomes out of stock or back in stock or is being enabled/disabled, it might influence how the parent product is being displayed on the frontend.

Does that answer your question @madonzy?

**Comment by madonzy** [comment]
@hostep not really. When child products become out of stock or enabled/disabled then child product cache tags are purged, and that means the cache of the category page will be purged. There's nothing to do with parent product in your examples.

**Comment by hostep** [comment]
I think I start to see your point, trying to fetch the parent products on the frontend makes not so much sense. But the `afterGetIdentities` method is also being called when a product is saved in the backoffice for example, and then it starts to make sense that for both the child and parent the caches are being purged, correct?

So if there is a performance problem here, it's that [these lines](https://github.com/magento/magento2/blob/a94527346fe0f9488a71248fae91b7a2e48b00a5/app/code/Magento/ConfigurableProduct/Model/Plugin/ProductIdentitiesExtender.php#L57-L64) should not be executed on the frontend, am I understanding this correctly?

**Comment by madonzy** [comment]
@hostep yeap. I think just disabling the plugin in the frontend scope di.xml will be a clean solution.

**Comment by engcom-Hotel** [comment]
@hostep thanks for helping us in understanding the issue.

Hello @madonzy,

We have looked into the code written [here](https://github.com/magento/magento2/blob/a94527346fe0f9488a71248fae91b7a2e48b00a5/app/code/Magento/ConfigurableProduct/Model/Plugin/ProductIdentitiesExtender.php#L57-L62) and tried to reproduce the issue by the following way:
1. Have some code changes to debug the issue as follows in the file named `ProductIdentitiesExtender.php` in line no.: 58:

```
$parentsidsbychild = $this->getParentIdsByChild($subject->getId());
        foreach ($parentsidsbychild as $parentId) {
            $parentProduct = $this->productRepository->getById($parentId);
            $parentProductsIdentities[] = $parentProduct->getIdentities();
        }
```
 
2. Search for the configurable product from Frontend i.e. `Chaz Kangeroo Hoodie`
3. Add a breakpoint in line no. 59 to check the value for the `$parentsidsbychild` variable, made in step 1.
4. It has been observed that code has not reached in `foreach` loop because the value of `$parentsidsbychild` is empty. 
5. Even if we have tried it for the product detail page.
6. But the method `afterGetIdentities` is calling every time.

In the next scenario we have tried to check the performance of the page by enabling and disabling the plugin and we have seen the below difference in them. Please have a look at the below screenshot:
### With Plugin:
<img width="1436" alt="Screenshot 2022-06-06 at 5 07 22 PM" src="https://user-images.githubusercontent.com/51681618/172165200-e214fd1a-67e6-42dc-bd46-d9beb28ba3ff.png">

### Without Plugin
<img width="1436" alt="Screenshot 2022-06-06 at 6 33 56 PM" src="https://user-images.githubusercontent.com/51681618/172166280-a276c760-4741-4cd4-9227-cc76f541b039.png">

I just need to confirm if we are on the correct path or if we can reproduce the issue in some other way.

Thanks

**Comment by madonzy** [comment]
@engcom-Hotel first of all it's totally strange that you did not get the parent product id from the child product. The second is that you measure not only the request but everything that is not needed :) 

Make sure that `getParentIdsByChild` is returning you the value and then got to the browser and test only document request time. Go to Network -> Document ->  Select main loaded document -> Timing

**Comment by engcom-Hotel** [comment]
Thanks @madonzy for the quick response. Let me try this and get back to you. :-)

**Comment by engcom-Hotel** [comment]
Hello @madonzy,

We have tried to reproduce the issue by searching for a configurable product in the frontend like `Chaz Kangeroo Hoodie-XS-Black` but we are always getting the Parent product id in the frontend, not the child product id. Please have a look at the below screenshot for the reference:

![image](https://user-images.githubusercontent.com/51681618/172360990-51e751f7-77c1-486f-b373-381677a78a72.png)

Instead, If I can get the Child id i.e. `53` in this case, I think we can get the Parent id. Can you please suggest the product for which we can get the child product in the frontend?

Thanks

**Comment by engcom-Hotel** [comment]
Hello @hostep,

I have tried this as well. But still, the issue is not reproducible for us. We have followed the below steps:
1. Create a configurable product with one configuration:
![image](https://user-images.githubusercontent.com/51681618/172651153-ac043f9b-138c-4fc6-b9af-95d4a71a5e0b.png)

2. Edit the configured product and mark the visibility as `catalog, search`:
![image](https://user-images.githubusercontent.com/51681618/172651521-c66bc4c6-b23c-435e-8f2c-43951fbc1ca8.png)

3. Cleared the cache
4. Check in the frontend, but for us, the function `afterGetIdentities` is only called for the parent product but not for the child product. For us the parent product id is `2049` and the child product id is `2050`. Please refer to the screencast for reference:

https://user-images.githubusercontent.com/51681618/172653719-f6e3ca66-160f-4025-a73b-a5f60bc696e8.mp4

Let us know in case we have missed anything.

Thanks

**Comment by hostep** [comment]
Hmm, I'm getting confused, just tried understanding [the code](https://github.com/magento/magento2/blob/a8543fed035d85c667195e95f69476d51f98381b/app/code/Magento/ConfigurableProduct/Model/Plugin/ProductIdentitiesExtender.php#L52-L65) again:
- So the `if` checks if a product if of type `configurable` or something else, in case of something else, it just returns the current `$identities`
- In case it's of type `configurable` the code continues and tries to find the parents of this configurable and then appends  their identities

So uhm, I'm confused now, since when can configurable products in Magento have parents?

@madonzy: what do you make of this? Do you have a case where a configurable product has parent products somehow? Or how were you able to trigger this code?

**Comment by madonzy** [comment]
There's a method that is called on product listing: \Magento\Catalog\Block\Product\ListProduct::getIdentities

There is a loop you can see fetching the identities of all products and merging it. Just make sure this method is called and all product are loaded.

**Comment by hostep** [comment]
Yes, but that's expected I think, otherwise the caching system in Magento can't invalidate such pages in case one of the products displayed on them gets changed.
Also, those products get loaded anyway to be able to display them, so it shouldn't really cause performance problems.

**Comment by hostep** [comment]
That's really strange, because [the return in these three lines of code](https://github.com/magento/magento2/blob/a8543fed035d85c667195e95f69476d51f98381b/app/code/Magento/ConfigurableProduct/Model/Plugin/ProductIdentitiesExtender.php#L54-L56) will trigger when it's a simple product. So this should not cause performance problems.

The code further in the method will only trigger when a configurable product is being loaded on the frontend and will try to load parents of the configurable product, which makes not much sense to me.

Have you already debugged this piece of a code a bit yourself? Does the code make sense why and how it triggers in your specific case?

**Comment by madonzy** [comment]
@hostep first of all, that is a new version, there's no such if in 2.3.6 for example https://github.com/magento/magento2/blob/2.3.6/app/code/Magento/ConfigurableProduct/Model/Plugin/ProductIdentitiesExtender.php

Also, I do not really understand the sense of that if statement... If the product is configurable then it will never have a parent product,(method getParentIdsByChild will always return an empty array)  so what is the purpose of that plugin?

**Comment by hostep** [comment]
Ah this is the first time you mention your Magento version (please next time do this in the opening post, it will make issue handling faster 😉 )

Indeed, the code in 2.3.6 is very different, it was changed significantly in https://github.com/magento/magento2/pull/30746 (which got included in Magento 2.4.2 and higher)

As to why the plugin still exists, I have no idea, either it can be thrown away, or a bug was introduced in https://github.com/magento/magento2/pull/30746, it's hard to say.

@engcom-Echo do you remember something about that Pull Request?

**Comment by engcom-Hotel** [comment]
Hello @madonzy,

It seems the issue has been fixed in the latest development branch i.e. 2.4-develop. Shall we close this issue?

Thanks

**Comment by madonzy** [comment]
@engcom-Hotel shell we remove that plugin at so?

**Comment by hostep** [comment]
There's a discussion about this going on [in Slack](https://magentocommeng.slack.com/archives/CBSL1DF8B/p1655217428425559):

<img width="1126" alt="Screenshot 2022-06-17 at 17 57 42" src="https://user-images.githubusercontent.com/85479/174334323-e0af9f99-47e0-4c6a-8b4c-9c8c75d2cd4a.png">

So it's being looked into I hope 🙂

**Comment by engcom-Hotel** [comment]
Hello @madonzy,

We are confirming this issue as per the below comments:

https://github.com/magento/magento2/issues/35546#issuecomment-1150058317
https://github.com/magento/magento2/issues/35546#issuecomment-1155287613

We can remove this plugin from here as it is not in use.

Thanks

**Comment by aholovan** [comment]
It's likely the same funny code in "add_bundle_parent_identities" plugin

**Comment by hostep** [comment]
https://github.com/magento/magento2/pull/36529 got merged yesterday, it changes some logic in the `ProductIdentitiesExtender::afterGetIdentities` method. I haven't tried to understand it yet, because it's quite complicated, but maybe it resolves this issue here @madonzy ?

**Comment by codedge** [comment]
@hostep @madonzy I was trying to remove the plugin via custom module `di.xml`

```xml
<config  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
    <type name="Magento\Catalog\Model\Product">
        <plugin name="product_identities_extender" disabled="true" />
    </type>
</config>
```

My module has these sequence

```
<module name="Codedge_CatalogSeo" setup_version="0.1.0">
        <sequence>
            <module name="Magento_Catalog"/>
            <module name="Magento_Theme"/>
            <module name="Magento_ConfigurableProduct"/>
        </sequence>
    </module>
```

but the plugin is still enabled if I check with `php bin/magento dev:di:info Magento\\Catalog\\Model\\Product`

Any idea?

**Comment by lbajsarowicz** [comment]
I checked ProductIdentitiesExtender.php and it is unchanged from what's described here; the question in this issue is architectural (why the plugin extends identities this way) rather than a defect. Can you clarify what behavior you'd expect instead, so it's clear whether this needs a code change or is working as designed?
